File: python/384-ShuffleAnArray/main.py
from typing import List
import random
import unittest

class Solution:

    def __init__(self, nums: List[int]):
        self.original_numbers = nums
        self.seed = 2166136261
        

    def reset(self) -> List[int]:
        return self.original_numbers
        

    def shuffle(self) -> List[int]:
        r = random.Random(self.seed)
        self.seed += 16777619

        pack = self.original_numbers.copy()
        r.shuffle(pack)

        return pack

class Tests(unittest.TestCase):
    def test_example1(self):
        solution = Solution([-6,10,184])
        print(solution.reset())
        print(solution.shuffle())
        print(solution.reset())
        print(solution.shuffle())
        print(solution.reset())
        print(solution.shuffle())


if __name__ == '__main__':
    unittest.main()

File: elixir/384-ShuffleAnArray/main.exs
defmodule Solution do
  @name __MODULE__

  defp start(nums) do
    Agent.start_link(fn -> nums end, name: @name)
  end

  defp restart(nums) do
    Agent.stop(@name)
    start(nums)
  end

  @spec init_(nums :: [integer]) :: any
  def init_(nums) do
    case start(nums) do
      {:ok, _} -> nil
      {:error, {:already_started, _}} -> restart(nums)
    end
  end

  @spec reset() :: [integer]
  def reset() do
    Agent.get_and_update(@name, fn nums -> {nums, nums} end)
  end

  @spec shuffle() :: [integer]
  def shuffle() do
    Agent.get_and_update @name, fn nums ->
      {Enum.shuffle(nums), nums}
    end
  end
end

defmodule Tests do
  def example1 do
    Solution.init_([1,2,3])
    IO.inspect(Solution.shuffle())
    IO.inspect(Solution.reset())
    IO.inspect(Solution.shuffle())

    Solution.init_([1,2])
    IO.inspect(Solution.shuffle())
    IO.inspect(Solution.reset())
    IO.inspect(Solution.shuffle())
  end
end

Tests.example1